### PR TITLE
Bugfix get protocol name logical.

### PR DESCRIPTION
**Pull request by wangfakang**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/stream/xprotocol/conn.go [hunk 1]
```diff
@@ -200,7 +200,13 @@ func (sc *streamConn) Dispatch(buf types.IoBuffer) {
 }
 
 func (sc *streamConn) Protocol() types.ProtocolName {
-	return protocol.Xprotocol
+	// it should priority return real protocol name
+	// ContextKeyDownStreamProtocol depend it
+	if sc.protocol != nil {
```
Comment (nejisama) [inline pkg/stream/xprotocol/conn.go:205]: XProtocol的场景，有ContextSubProtocol 表明子协议，这个改动可能会影响较多的逻辑，需要确定是否必须这样改动 @neverhook

Comment (wangfakang) [reply pkg/stream/xprotocol/conn.go:205]: 目前 MOSN 中有三处依赖 `Protocol()` 这个接口，为了获取其真实协议名称。 
https://github.com/mosn/mosn/blob/master/pkg/proxy/downstream.go#L640
https://github.com/mosn/mosn/blob/master/pkg/proxy/downstream.go#L136
https://github.com/mosn/mosn/blob/master/pkg/proxy/downstream.go#L129

其中 `convertProtocol()` 会获取up 和 down的协议，如果此时 down 是 `http`，up 是 `X` 也会做没有必要的转换。
```
func (s *downStream) convertProtocol() (dp, up types.ProtocolName) {
        dp = s.getDownstreamProtocol()
        up = s.getUpstreamProtocol()
        return
}

func (s *downStream) getDownstreamProtocol() (prot types.ProtocolName) {
        if s.proxy.serverStreamConn == nil {
                prot = types.ProtocolName(s.proxy.config.DownstreamProtocol)
        } else {
                prot = s.proxy.serverStreamConn.Protocol()
        }
        return prot
}


```

File: pkg/proxy/downstream.go [hunk 2]
```diff
@@ -650,8 +650,8 @@ func (s *downStream) getUpstreamProtocol() (currentProtocol types.ProtocolName)
 		configProtocol = s.route.RouteRule().UpstreamProtocol()
 	}
 
-	// Auto means same as downstream protocol
-	if configProtocol == string(protocol.Auto) {
+	// Auto or X means same as downstream protocol
+	if configProtocol == string(protocol.Auto) || configProtocol == string(protocol.Xprotocol) {
```
Comment (nejisama) [inline pkg/proxy/downstream.go:654]: 这个改动以后，如果down是HTTP，up是X，会被强制改成HTTP，这个行为是错的。
没有提供默认的协议转换，不代表不支持协议转换

File: pkg/stream/xprotocol/conn.go (hunk 1 above)
Comment (wangfakang) [reply pkg/stream/xprotocol/conn.go:205]: 讨论了下，目前Xprotocol的connpool有依赖X，暂时使用的地方先使用其他的方式修复。

**Pull request merged**